Softeq.XToolkit.Common/Commands/AsyncCommandGeneric.cs

```diff
@@ -0,0 +1,140 @@
+        /// </param>
+        public AsyncCommand(Func<T, Task> execute)
+            : this(execute, _ => true)
+        {
+        }
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="AsyncCommand{T}"/> class.
+        /// </summary>
+        /// <param name="execute">
+        ///     The Function executed when Execute or ExecuteAsync is called.
+        ///     This does not check canExecute before executing and will execute even if canExecute is false.
+        /// </param>
+        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
+        public AsyncCommand(Func<T, Task> execute, Action<Exception> onException)
+            : this(execute, _ => true, onException)
+        {
+        }
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="AsyncCommand{T}"/> class.
+        /// </summary>
+        /// <param name="execute">
+        ///     The Function executed when Execute or ExecuteAsync is called.
+        ///     This does not check canExecute before executing and will execute even if canExecute is false.
+        /// </param>
+        /// <param name="canExecute">The Function that verifies whether or not AsyncCommand should execute.</param>
+        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
+        public AsyncCommand(Func<T, Task> execute, Func<bool> canExecute, Action<Exception>? onException = null)
+            : this(execute, _ => canExecute?.Invoke() ?? true, onException)
+        {
+        }
+
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="AsyncCommand{T}"/> class.
+        /// </summary>
+        /// <param name="execute">
+        ///     The Function executed when Execute or ExecuteAsync is called.
+        ///     This does not check canExecute before executing and will execute even if canExecute is false.
+        /// </param>
+        /// <param name="canExecute">The Function that verifies whether or not AsyncCommand should execute.</param>
+        /// <param name="onException">If an exception is thrown in the Task, <c>onException</c> will execute.</param>
+        public AsyncCommand(
+            Func<T, Task> execute,
+            Func<object?, bool>? canExecute = null,
+            Action<Exception>? onException = null)
+            : base(canExecute)
+        {
+            _execute = execute;
+            _onException = onException;
+        }
+
+        /// <inheritdoc cref="AsyncCommandBase.CanExecute"/>
+        public bool CanExecute(T parameter)
+        {
+            return CanExecute(parameter as object);
+        }
+
+        /// <inheritdoc cref="AsyncCommand.Execute"/>
+        /// <typeparam name="T">Type of parameter.</typeparam>
+        public void Execute(T parameter)
+        {
+            ExecuteInternal(parameter);
```

Comment:
I think better to maintain and extend when it will be in one method without any casts.

Reply:
Sorry, but I can't quite understand what you mean. Can you explain, please?
In this case compiler make all the type checks - all possible errors would be detected in compilation time. There's absolutely no possibility, that `T parameter` is invalid inside this method - that's why I suggest to call `ExecuteAsync(parameter)` here directly. That would also mean that `ExecuteInternal` could be simplified - it doesn't need to be generic, as it will be called only for `object?` type from `Execute(object? parameter)`

Reply:
It was also done for cases when we have T as nullable struct.

Added tests: https://github.com/Softeq/XToolkit.WhiteLabel/pull/293/commits/df31bbd53200cc300400a91f12e513ec60c11d85

Reply:
Regarding [this test](https://github.com/Softeq/XToolkit.WhiteLabel/commit/df31bbd53200cc300400a91f12e513ec60c11d85#diff-e0211544c1e142eb75483931f369ac73R147):
is it really intended behavior? Seems valid to me to call `Action<int?>` with parameter null, if it is actually of type `int?`. It is definitely not intuitive to get an exception in this case.

Reply:
Agree! Fixed